java-checks/src/main/java/org/sonar/java/checks/FunctionalInterfacesSpecialisedCheck.java

```diff
@@ -0,0 +1,257 @@
+    } else if (firstArgument.originalName.equals(secondArgument.originalName)) {
+      return !args.get(0).is(Tree.Kind.PARAMETERIZED_TYPE) ? (UNARY_OPERATOR + firstArgument.originalName + ">") : (UNARY_OPERATOR + firstArgument.originalName);
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private static String handleBiFunctionInterface(TypeArguments args) {
+    ParameterTypeNameAndTreeType firstArgument = new ParameterTypeNameAndTreeType(args.get(0));
+    ParameterTypeNameAndTreeType secondArgument = new ParameterTypeNameAndTreeType(args.get(1));
+    ParameterTypeNameAndTreeType thirdArgument = new ParameterTypeNameAndTreeType(args.get(2));
+    return (firstArgument.originalName.equals(secondArgument.originalName) &&
+      firstArgument.originalName.equals(thirdArgument.originalName)) ? ("BinaryOperator<" + firstArgument.originalName + ">") : null;
+  }
+
+  @CheckForNull
+  private static String handleBiConsumerInterface(TypeArguments args) {
+    return "Obj" + returnStringFromJavaObject(((IdentifierTree) args.get(1)).symbolType()) + "Consumer";
+  }
+
+  @CheckForNull
+  private static String returnStringFromJavaObject(Type argType) {
+    if (argType.is("java.lang.Integer")) {
+      return "Int";
+    } else if (argType.is("java.lang.Boolean") || argType.is("java.lang.Double") || argType.is("java.lang.Long")) {
+      return argType.name();
+    } else {
+      return null;
+    }
+  }
+
+  @CheckForNull
+  private static TypeArguments interfaceParameters(TypeTree interfaceType) {
+    TypeArguments args = null;
+    if (interfaceType.is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      ParameterizedTypeTree parameters = (ParameterizedTypeTree) interfaceType;
+      args = parameters.typeArguments();
+    }
+    return args;
+  }
+
+  private void report(Tree tree, String reportString) {
+    if (tree.is(Tree.Kind.CLASS)) {
+      reportIssue(((ClassTree) tree).superInterfaces(), REPORT_STRING + reportString + "'");
+    } else {
+      reportIssue(tree, REPORT_STRING + reportString + "'");
+    }
+  }
+
+  private static class ParameterTypeNameAndTreeType {
+    Type paramSymbol;
+    String paramTypeName;
+    String originalName;
+
+    public ParameterTypeNameAndTreeType(Tree tree) {
+      switch (tree.kind()) {
+        case IDENTIFIER:
+          IdentifierTree idTree = (IdentifierTree) tree;
+          this.paramSymbol = idTree.symbolType();
+          this.paramTypeName = returnStringFromJavaObject(this.paramSymbol);
+          this.originalName = idTree.name();
+          break;
+        case MEMBER_SELECT:
+          MemberSelectExpressionTree memberTree = (MemberSelectExpressionTree) tree;
+          this.paramSymbol = memberTree.identifier().symbolType();
+          this.paramTypeName = returnStringFromJavaObject(this.paramSymbol);
+          this.originalName = ((IdentifierTree) memberTree.expression()).name() + "." + memberTree.identifier().name();
```

Comment:
This won't work and crash in the following case: `a.b.c.Foo`. Just use the last identifier.